clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1261,6 +1270,15 @@ class SyclKernelBodyCreator
   using SyclKernelFieldHandler::leaveStruct;
 };
 
+/// Tells whether given type is a sycl accessor type which must be mapped to
+/// ESIMD buffer accessor.
+static bool isESIMDAccessor(const CXXRecordDecl *AccTy) {
+  // Cooperate with the SYCL API, which defines this method iff the accessor is
+  // ESIMD accessor.
+  assert(Util::isSyclAccessorType(AccTy) && "sycl::accessor expected");
+  return getMethodByName(AccTy, "ESIMDBufferAccessorMarker") != nullptr;
```

Comment:
I don't get why we have to distinguish ESIMD accessors from default ones if user cannot specify that accessor is ESIMD accessor and it is not possible to get a program with ESIMD and default accessors at the same time.

Reply:
There will be corresponding runtime patch, which makes it clear why, + provides more detailed comments.

In short, for accessors in ESIMD kernels memory object creation is different - it is not `clCreateBuffer`

Reply:
Okay, in this case there could be difference in process of setting kernel arguments, that is why you need to change descriptor in the integration header.
But I still don't get this searching for internal method. Why we don't threat all accessors as ESIMD ones when ESIMD extension is enabled?

Reply:
Because some day we want to mix SIMD and normal kernels.

Reply:
I'm a bit concerned about having additional handling for ESIMD accessors. I think, we were moving to unification of kernel arguments handling. This stuff just breaks this direction. Is it possible to elaborate what is the difference in RT handling between SIMD accessors and default ones?

Reply:
I explained above:
> In short, for accessors in ESIMD kernels memory object creation is different - it is not clCreateBuffer

that's the only difference.

Reply:
CUDA and L0 back-ends doesn't use `clCreateBuffer` to allocate the memory and still they don't require special accessors. I don't follow why ESIMD kernels memory object creation requires compiler changes.

Reply:
> CUDA and L0 back-ends doesn't use clCreateBuffer to allocate the memory and still they don't require special accessors. I don't follow why ESIMD kernels memory object creation requires compiler changes.

This is a good example, which can help clarify.

CUDA and L0 are separate back-ends, as you correctly mentioned. What I need here is to distinguish usual accessors and ESIMD accessors within the same backend - be it L0 or OpenCL. I see that L0 is finally integrated into open-source, and I'll need to adjust the changes, but their sense will be the same - different sequence of PI calls will be needed to create memory object for the ESIMD accessor.

I could tag a kernel with "I am SIMD kernel" info, then diverge accessor handling based on that. But my estimation was that that would be much more intrusive and unsafe change to the runtime. I can add a TODO naming this alternative.

Reply:
I explained my concerns above. We were moving to unification of kernel arguments generation, so we can improve detection and handling of special SYCL types (i.e. accessors, sampler and streams) by using an attribute. This approach was discussed during upstreaming WG meeting and we agreed that it is better than what we have now. Adding new accessor type doesn't seem like the change compatible with this direction, so I would prefer tagging kernels. Although I don't have enough expertise in RT to evaluate how intrusive and unsafe change to the runtime. @bader, WDYT?